src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp

```diff
@@ -1639,33 +1639,67 @@ void C2_MacroAssembler::minmax_fp_v(VectorRegister dst, VectorRegister src1, Vec
   vfadd_vv(dst, src2, src2, Assembler::v0_t);
 }
 
+// Set dst to NaN if any NaN input.
+// The destination vector register elements corresponding to masked-off elements
+// are handled with a mask-undisturbed policy.
+void C2_MacroAssembler::minmax_fp_masked_v(VectorRegister dst, VectorRegister src1, VectorRegister src2,
+                                           VectorRegister vmask, int vector_length, VectorRegister tmp1,
+                                           VectorRegister tmp2, bool is_double, bool is_min) {
+  assert_different_registers(src1, src2, tmp1, tmp2);
+  vsetvli_helper(is_double ? T_DOUBLE : T_FLOAT, vector_length);
+
+  // Check vector elements of src1 and src2 for NaN.
+  vmfeq_vv(tmp1, src1, src1);
+  vmfeq_vv(tmp2, src2, src2);
+
+  vmandn_mm(v0, vmask, tmp1);
+  vfadd_vv(dst, src1, src1, Assembler::v0_t);
+  vmandn_mm(v0, vmask, tmp2);
+  vfadd_vv(dst, src2, src2, Assembler::v0_t);
+
+  vmand_mm(tmp2, tmp1, tmp2);
+  vmand_mm(v0, vmask, tmp2);
+  is_min ? vfmin_vv(dst, src1, src2, Assembler::v0_t)
+         : vfmax_vv(dst, src1, src2, Assembler::v0_t);
+}
+
 // Set dst to NaN if any NaN input.
 void C2_MacroAssembler::reduce_minmax_fp_v(FloatRegister dst,
                                            FloatRegister src1, VectorRegister src2,
                                            VectorRegister tmp1, VectorRegister tmp2,
-                                           bool is_double, bool is_min, int vector_length) {
+                                           bool is_double, bool is_min, int vector_length, VectorMask vm) {
   assert_different_registers(src2, tmp1, tmp2);
 
-  Label L_done, L_NaN;
+  Label L_done, L_done_check, L_NaN_v, L_NaN_f;
+  // If src1 is NaN, set dst to src1 directly.
+  is_double ? feq_d(t0, src1, src1)
+            : feq_s(t0, src1, src1);
+  beqz(t0, L_NaN_f);
+
   vsetvli_helper(is_double ? T_DOUBLE : T_FLOAT, vector_length);
   vfmv_s_f(tmp2, src1);
 
-  is_min ? vfredmin_vs(tmp1, src2, tmp2)
-         : vfredmax_vs(tmp1, src2, tmp2);
+  is_min ? vfredmin_vs(tmp1, src2, tmp2, vm)
+         : vfredmax_vs(tmp1, src2, tmp2, vm);
```

Comment:
Suggestion: put the result of reduction in `dst` with `vfmv_f_s(dst, tmp1)` here and save the `j(L_done_check)` at line 1695.

Reply:
Fixed.